relations/src/shielder/circuit_utils.rs

```diff
@@ -0,0 +1,69 @@
+use core::borrow::Borrow;
+#[cfg(feature = "std")]
+use std::fmt::{Display, Formatter};
+
+use ark_r1cs_std::{
+    alloc::{AllocVar, AllocationMode},
+    boolean::Boolean,
+    R1CSVar,
+};
+use ark_relations::r1cs::{Namespace, SynthesisError};
+
+use crate::CircuitField;
+
+#[derive(Clone, Debug)]
+pub(super) struct PathShapeVar {
+    shape: Vec<Boolean<CircuitField>>,
+}
+
+#[cfg(feature = "std")]
+impl Display for PathShapeVar {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{:?}",
+            self.shape
+                .iter()
+                .map(|b| b.value().map(|boo| if boo { "left" } else { "right" }))
+                .collect::<Vec<_>>()
+        )
+    }
+}
+
+impl PathShapeVar {
+    pub(super) fn len(&self) -> usize {
+        self.shape.len()
+    }
+
+    pub(super) fn at(&self, i: usize) -> &Boolean<CircuitField> {
+        &self.shape[i]
+    }
+}
+
+impl AllocVar<(u8, Result<u64, SynthesisError>), CircuitField> for PathShapeVar {
+    fn new_variable<T: Borrow<(u8, Result<u64, SynthesisError>)>>(
+        cs: impl Into<Namespace<CircuitField>>,
+        f: impl FnOnce() -> Result<T, SynthesisError>,
+        mode: AllocationMode,
+    ) -> Result<Self, SynthesisError> {
+        let ns = cs.into();
+        let cs = ns.cs();
+
+        let mut shape = vec![];
+
+        let (path_length, maybe_leaf_index) = *f()?.borrow();
+
+        for i in 0..path_length {
+            shape.push(Boolean::new_variable(
+                cs.clone(),
+                || {
+                    let current_index = maybe_leaf_index? / (1 << i);
```

Comment:
Why does the `maybe_leaf_index` have to be unwrapped inside the `f` of `Boolean_new_variable` and cannot be unwrapped immediately in line 54?

Reply:
yes, that's the tricky part :) both for generating keys (`RelationWithoutInput`) and for generating proof (`RelationWithFullInput`) we know `max_path_length`; however, only during proof generation we actually need `leaf_index` **and only during proof generation this closure (for `new_variable`) will be actually triggered**

Reply:
I see - so the circuit is the same for generating keys and proof but only in the second case the `leaf_index` is known and used.

Reply:
exactly